### PR TITLE
Add Brocfile.ts TypeScript support

### PR DESCRIPTION
**Pull request by oligriffiths**
This PR adds support for loading a typescript Brocfile and running it through `ts-node`.
If one provides a `Brocfile.ts`, install `ts-node` and `typescript`, it will be auto picked up and run through ts-node.
An optional `tsconfig.json` can exist alongside the `Brocfile.ts`, however defaults taken from https://github.com/stefanpenner/fs-tree-diff/pull/89/files#diff-e5e546dd2eb0351f813d63d1b39dbc48 are provided

I think I've provided a type definition for the `BrocfileOptions`, but I may have gotten this wrong, plz inform.
Tests added.

### CODE REVIEW

**Comment by oligriffiths** [comment]
@stefanpenner @rwjblue have you had a chance to check this out? Am I on the right path?

**Comment by oligriffiths** [comment]
@chriskrycho Would you be interested in having a look at this given your work on ember-typescript?

**Comment by oligriffiths** [comment]
@mike-north Would you be interested in having a look at this given your work on ember-typescript?

File: lib/load_brocfile.js [hunk 1]
```diff
@@ -4,23 +4,76 @@ const path = require('path');
...
+          moduleResolution: 'node',
+          module: 'commonjs',
+          target: 'es5',
+          baseUrl: '.',
+        },
+      };
+    }
+
+    // Ensure any previous .tsx? extensions are removed
+    delete require.extensions['.ts']; // eslint-disable-line node/no-deprecated-api
```
Comment (dfreeman) [inline lib/load_brocfile.js]: If someone has already set up `ts-node` (or otherwise decided to handle `.ts` files), shouldn't we defer to that instead of blowing it away?

File: lib/load_brocfile.js [hunk 2]
```diff
@@ -4,23 +4,76 @@ const path = require('path');
...
+    const configPath = findup('tsconfig.json', {
+      nocase: true,
+    });
+
+    let options;
+
+    if (configPath) {
+      options = { project: configPath };
+    } else {
+      options = {
```
Comment (dfreeman) [inline lib/load_brocfile.js]: If no `tsconfig.json` file is found (or maybe even if it is), would it make sense to use `ts-node` in `transpileOnly` mode? It's a bit faster since it skips typechecking, and you don't risk enabling config options that will break the user.

File: lib/load_brocfile.js (hunk 1 above)
Comment (oligriffiths) [reply lib/load_brocfile.js]: yeah I realise this was just for tests, so I will add this to the tests instead

File: lib/load_brocfile.js (hunk 2 above)
Comment (oligriffiths) [reply lib/load_brocfile.js]: honestly not sure, happy to take advice on this for recommended default (can always be changed with tsconfig)

File: lib/load_brocfile.js [hunk 3]
```diff
@@ -4,23 +4,71 @@ const path = require('path');
...
+        // Define default compilerOptions
+        compilerOptions: {
```
Comment (oligriffiths) [inline lib/load_brocfile.js]: Looking for a steer on these defaults

Comment (chriskrycho) [reply lib/load_brocfile.js]: These seem reasonable; the only one I'd consider changing is `target` – I'd set it to whatever the current Node LTS supports (which IIRC is ES2015? Might already be 2016).

Comment (oligriffiths) [reply lib/load_brocfile.js]: I think it should be node 6, which is what's in the package.json as the min node version. Agree?

Comment (chriskrycho) [reply lib/load_brocfile.js]: Seems 👍 to me!

File: lib/load_brocfile.js [hunk 4]
```diff
@@ -4,23 +4,71 @@ const path = require('path');
...
+ * Require a brocfile via either ESM or TypeScript
+ *
+ * @param {String} brocfilePath The path to the brocfile
+ * @returns {*}
+ */
+function requireBrocfile(brocfilePath) {
+  let brocfile;
+
+  if (brocfilePath.match(/\.ts$/)) {
+    const configPath = findup('tsconfig.json', {
```
Comment (rwjblue) [inline lib/load_brocfile.js]: Why is this needed at all?

https://github.com/TypeStrong/ts-node#loading-tsconfigjson suggests that it will find the tsconfig.json just fine on its own, no?

File: lib/load_brocfile.js (hunk 2 above)
Comment (mike-north) [reply lib/load_brocfile.js]: typescript already has some defaults it falls back to in the absence of a `tsconfig.json`. We may want to consider remaining consistent with that, in order to avoid introducing ts-node as a dependency, and confusing those who expect "normal" behavior from config-less typescript compilation

File: package.json [hunk 5]
```diff
@@ -43,6 +44,8 @@
     "sane": "^4.0.0",
     "tmp": "0.0.33",
     "tree-sync": "^1.2.2",
+    "ts-node": "^8.0.2",
+    "typescript": "^3.3.3",
```
Comment (mike-north) [inline package.json]: you'll certainly want to make this `~3.3.3`, since typescript does not follow semver, and regularly includes breaking changes in "middle-number" releases

File: lib/load_brocfile.js (hunk 4 above)
Comment (stefanpenner) [reply lib/load_brocfile.js]: @oligriffiths thoughts?

File: lib/load_brocfile.js [hunk 6]
```diff
@@ -4,23 +4,84 @@ const path = require('path');
...
+          strict: true,
+          moduleResolution: 'node',
+          module: 'commonjs',
+          target: 'ES2015', // Node 6 per package.json
+          baseUrl: '.',
+        },
+      };
+    }
+
+    try {
```
Comment (stefanpenner) [inline lib/load_brocfile.js]: @rwjblue I think this addresses your concerns, can you confirm/deny?

File: package.json (hunk 5 above)
Comment (oligriffiths) [reply package.json]: Dependency removed

File: lib/load_brocfile.js (hunk 4 above)
Comment (oligriffiths) [reply lib/load_brocfile.js]: Well, the point of this is to ensure that we either use the tsconfig or we use the defaults, but not both. node-ts overwrites the tsconfig options with the passed in config https://github.com/TypeStrong/ts-node/blob/2609663bf3a5b14d76c0dfe989ceba67e68d5efa/src/index.ts#L486

**Comment by oligriffiths** [comment]
Ok, so here's an update.
I've removed the `ts-node` and `typescript` dependencies, and now throw an exception if they're not installed, prompting the user to install them.
However, I have some questions on how we want `ts-node` to work by default, and if we should provide default options.

1. Should type definitions be defined for imported modules? For example, doing `import merge from 'broccoli-merge-trees'`, TS complains that there's no type information for this, and it's defaulting to `any` type. `noImplicitAny: false` is required to fix this
2. How should importing from CommonJS modules work? This is affected by `esModuleInterop: true` config option if we want it to work out of the box for `import merge from 'broccoli-merge-trees'`
3. Should we allow importing a JS file from a TS file? If so, `allowJs: true` is required.
4. if `allowJs: true` is used, the tests fail when there is more than 1 test that loads a TS brocfile (a single one passes, multiple fail):
```
     TSError: ⨯ Unable to compile TypeScript:
Brocfile.ts(2,24): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
Brocfile.ts(2,32): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
Brocfile.ts(5,23): error TS2304: Cannot find name 'exports'.
Brocfile.ts(6,43): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i @types/node` and then add `node` to the types field in your tsconfig.
Brocfile.ts(8,1): error TS2304: Cannot find name 'exports'.
```

😞

**Comment by mike-north** [comment]
> Should type definitions be defined for imported modules? For example, doing import merge from 'broccoli-merge-trees', TS complains that there's no type information for this, and it's defaulting to any type. noImplicitAny: false is required to fix this. The downside of requiring this is that no broccoli plugins currently have type definitions, so this would be painful for early adopters.

this is a pretty general issue having to do with typescript. Consumers should either define their own types locally, PR types into the dependency (through a declaration file or a JS -> TS conversion) or add a package to DefinitelyTyped

> How should importing from CommonJS modules work? This is affected by esModuleInterop: true config option if we want it to work out of the box for import merge from 'broccoli-merge-trees'

the way to do this without forcing consumers to set this flag is
```ts
// import
import foo = require('foo');

// export
export = 'foo';
```

File: lib/load_brocfile.js (hunk 2 above)
Comment (oligriffiths) [reply lib/load_brocfile.js]: Removed the tsconfig loading, defaults to the project dir tsconfig.json anyway, and works out of the box.

File: lib/load_brocfile.js (hunk 4 above)
Comment (oligriffiths) [reply lib/load_brocfile.js]: Removed

File: README.md [hunk 7]
```diff
@@ -88,6 +88,30 @@ module.exports = (options) => {
...
+
+export default (options: BrocfileOptions) => {
+  // tree = ... assemble tree
+
+  // In production environment, minify the files
+  if (options.env === 'production') {
+    tree = minify(tree);
+  }
+
+  return tree';
```
Comment (stefanpenner) [inline README.md]: Some funky syntax here

**Comment by oligriffiths** [comment]
Thank you to everyone who contributed feedback on this PR 🎉
